test/browser/features/steps/browser_steps.rb

```diff
@@ -101,3 +101,37 @@
     And the payload has a valid sessions array
   }
 end
+
+Then("the event device ID is valid") do
+  if has_local_storage
+    step('the event "device.id" matches "^c[a-z0-9]{20,32}$"')
+  elsif can_run_javascript
+    $logger.info('Local storage is not supported in this browser, assuming device ID is null')
+    step('the event "device.id" is null')
+  else
+    # if there's no local storage and we can't run javascript from Selenium, we
+    # don't know if the device ID is null or not, so we can't assert against it
+    $logger.warn('Selenium cannot run JavaScript in this browser. Device ID will not be checked!')
+  end
+end
+
+Then("the event device ID is {string}") do |expected_id|
+  if has_local_storage
+    step("the event \"device.id\" equals \"#{expected_id}\"")
+  elsif can_run_javascript
+    $logger.info('Local storage is not supported in this browser, assuming device ID is null')
+    step('the event "device.id" is null')
+  else
+    # if there's no local storage and we can't run javascript from Selenium, we
+    # don't know if the device ID is null or not, so we can't assert against it
+    $logger.warn('Selenium cannot run JavaScript in this browser. Device ID will not be checked!')
+  end
+end
+
+When("I save {string} in local storage as {string}") do |key, value|
```

Comment:
Not having local storage available seems like grounds to skip this test instead ?

Reply:
it does, but as established, this check failing doesn't mean local storage is not available, on mobile devices it just means you can execture arbitrary JS over the selenium interface

Reply:
I changed this to use a new HTML page to setup local storage instead, so that we don't need to run javascript via Selenium to run this test

It now also skips if we know local storage isn't available (i.e. Safari 10)